### PR TITLE
Refactor TektonConfig Reconciler and Cleanup

### PR DESCRIPTION
**Pull request by sm43**
Fixes addon component deletion 

If clustertask and pipelines templates are disabled then
due to return statement it used to return early without
executing remaning code.

-----

 Updates TektonConfig Conditions and cleanup

This updates the tektonconfig conditions and removes
unnecessary code

----

 Removes deletion timestamp check in Autocreation of TektonConfig

If TektonConfig has a deletion stamp, it won't be deleted
unless the operator is up and running as the object has
finalizer. waiting for it to get deleted will stop operator
from running and object still won't be deleted.
hence removing the condition for checking deletion timestamp
and letting user to create a new instance of tektonconfig
after deletion.

Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go [hunk 1]
```diff
@@ -21,107 +21,128 @@ import (
...
+)
 
+var (
 	configCondSet = apis.NewLivingConditionSet(
-		DependenciesInstalled,
-		DeploymentsAvailable,
-		InstallSucceeded,
+		PreReconciler,
+		ComponentsReady,
+		PostReconciler,
```
Comment (savitaashture) [inline pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:32]: I assume these conditions are displayed on status and i think `PreReconciler` and `PostReconciler` words are more towards technical

May i know the reason behind removing old conditions

Comment (nikhil-thomas) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:32]: > PreReconciler and PostReconciler words are more towards technical

I agree with this. It would be great if we could rename it with something that indicates the objective/end result of the action. It doesn't has to be precise. Even `PreInstall` `PostInstall` will be better.

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go [hunk 2]
```diff
@@ -21,107 +21,128 @@ import (
...
-	_ TektonComponentStatus = (*TektonConfigStatus)(nil)
+const (
+	ComponentsReady apis.ConditionType = "ComponentsReady"
+)
 
+var (
 	configCondSet = apis.NewLivingConditionSet(
-		DependenciesInstalled,
-		DeploymentsAvailable,
-		InstallSucceeded,
```
Comment (nikhil-thomas) [inline pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:30]: shall we retain `InstallSucceeded` as the final condition. Then we could share that as the standard way to check whether all aspects of TektonConfig installation was completed successfully.

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go [hunk 3]
```diff
@@ -24,11 +24,6 @@ import (
 	duckv1 "knative.dev/pkg/apis/duck/v1"
 )
 
-var (
-	_ TektonComponent     = (*TektonConfig)(nil)
-	_ TektonComponentSpec = (*TektonConfigSpec)(nil)
-)
-
```
Comment (nikhil-thomas) [inline pkg/apis/operator/v1alpha1/tektonconfig_types.go:31]: what makes TektonConfig not implement the TektonComponent Interface.
If TetkonConfig does not implement TektonComponent Interface, will that restrict us in some way in tests or in extensions.

File: pkg/reconciler/openshift/tektonaddon/tektonaddon.go [hunk 4]
```diff
@@ -185,7 +185,9 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, ta *v1alpha1.TektonAddon
 		}
 	} else {
 		// if disabled then delete the installer Set if exist
-		return r.deleteInstallerSet(ctx, ta, clusterTaskInstallerSet)
+		if err := r.deleteInstallerSet(ctx, ta, clusterTaskInstallerSet); err != nil {
+			return err
+		}
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonaddon/tektonaddon.go:190]: nit: is this change really necessary.

File: pkg/reconciler/openshift/tektonaddon/tektonaddon.go [hunk 5]
```diff
@@ -207,7 +209,9 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, ta *v1alpha1.TektonAddon
 		}
 	} else {
 		// if disabled then delete the installer Set if exist
-		return r.deleteInstallerSet(ctx, ta, pipelinesTemplateInstallerSet)
+		if err := r.deleteInstallerSet(ctx, ta, pipelinesTemplateInstallerSet); err != nil {
+			return err
+		}
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonaddon/tektonaddon.go:214]: same 🧑‍💻

File: pkg/reconciler/kubernetes/tektonconfig/extension.go [hunk 6]
```diff
@@ -52,13 +51,7 @@ func (oe kubernetesExtension) PostReconcile(ctx context.Context, comp v1alpha1.T
...
 
 	if configInstance.Spec.Profile == common.ProfileLite || configInstance.Spec.Profile == common.ProfileBasic {
-		err := extension.TektonDashboardCRDelete(oe.operatorClientSet.OperatorV1alpha1().TektonDashboards(), common.DashboardResourceName)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
-			return err
-		}
+		return extension.TektonDashboardCRDelete(oe.operatorClientSet.OperatorV1alpha1().TektonDashboards(), common.DashboardResourceName)
```
Comment (nikhil-thomas) [inline pkg/reconciler/kubernetes/tektonconfig/extension.go:54]: i think, here the code should ignore notFound error.

if initially the profile was `LITE` and then it was changed to `BASIC`, then we will get a `NotFound` error for delete Dashboard, as Dashboard is not provided in either of the profiles.

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 1 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:32]: yeah `PreInstall` `PostInstall` will also be nice
also we do want to break down the component status in some way currently the components cond will show any error
in further interation we do want to find a better way for it

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 2 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:30]: currently i haven't added any cond for pruner .,. so kept a common `componentready` for all ..
we can look into it more
what would `InstallSucceeded` say?

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go (hunk 3 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:31]: we can enable it back but with some changes I have added a TODO in lifecycle for it
once we refactor al components (dashboard, result ) we can find common things and update TektonComponent
otherwise they would break

File: pkg/reconciler/openshift/tektonaddon/tektonaddon.go (hunk 4 above)
Comment (sm43) [reply pkg/reconciler/openshift/tektonaddon/tektonaddon.go:190]: yeah it is required, if clustertasks are disabled then rest of the code was not executed 
so we only return if there is some err while deleting

File: pkg/reconciler/openshift/tektonaddon/tektonaddon.go (hunk 5 above)
Comment (sm43) [reply pkg/reconciler/openshift/tektonaddon/tektonaddon.go:214]: same as above 😄

File: pkg/reconciler/kubernetes/tektonconfig/extension.go (hunk 6 above)
Comment (sm43) [reply pkg/reconciler/kubernetes/tektonconfig/extension.go:54]: actually the not found case is already handled inside `TektonDashboardCRDelete` so we don't need to explicitly handle it :)

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 2 above)
Comment (savitaashture) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:30]: `InstallSucceeded` shows the success of Operator installation basically success of TektonConfig installation

File: pkg/reconciler/kubernetes/tektonconfig/extension.go (hunk 6 above)
Comment (nikhil-thomas) [reply pkg/reconciler/kubernetes/tektonconfig/extension.go:54]: oh that is wonderful 🧑‍💻 👍

File: pkg/reconciler/openshift/tektonaddon/tektonaddon.go (hunk 4 above)
Comment (nikhil-thomas) [reply pkg/reconciler/openshift/tektonaddon/tektonaddon.go:190]: oh.. i see. it is about `not` returning nil. and not about returning error. 👍

File: pkg/reconciler/openshift/tektonconfig/extension.go [hunk 7]
```diff
@@ -30,45 +28,26 @@ import (
...
-	mflogger := zapr.NewLogger(logger.Named("manifestival").Desugar())
-	manifest, err := mf.ManifestFrom(mf.Slice{}, mf.UseClient(mfclient), mf.UseLogger(mflogger))
-	if err != nil {
-		logger.Fatalw("error creating initial manifest", zap.Error(err))
-	}
-	ext := openshiftExtension{
+	return openshiftExtension{
 		operatorClientSet: operatorclient.Get(ctx),
 		kubeClientSet:     kubeclient.Get(ctx),
-		manifest:          manifest,
```
Comment (nikhil-thomas) [inline pkg/reconciler/openshift/tektonconfig/extension.go:62]: 👍

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 2 above)
Comment (nikhil-thomas) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:30]: > currently i haven't added any cond for pruner .,. so kept a common componentready for all ..

ok.. copy that.
i recommend adding a final `InstallSucceed` so that QE and admins can use scripts for monitoring this one condition, no matter what how we add other intermediate conditions (or add more checks before setting `InstallSucceed: True`)

Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:30]: by default `Ready` condition is added by knative/pkg which gets true only when all other cond get true
so `InstallSucceeded` is diff from `Ready` ?

Comment (nikhil-thomas) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:30]: Oh.. great 👍 forgot about that.
That is enough.

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go (hunk 3 above)
Comment (nikhil-thomas) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:31]: could you add it as an issue in https://github.com/tektoncd/operator/issues

Comment (nikhil-thomas) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:31]: my concern is that, if we don't have this checks, we might end up loosing uniformity in methods of components in the future. It is not a deal breaker, but i would like a reminder to check this later.

**Comment by sm43** [comment]
/hold
to rename the conditions

File: pkg/apis/operator/v1alpha1/tektonaddon_types.go [hunk 8]
```diff
@@ -65,6 +67,17 @@ type TektonAddonStatus struct {
...
 
+// Addon defines the field to customize Addon component
+type Addon struct {
+	// Params is the list of params passed for Addon customization
+	// +optional
+	Params []Param `json:"params,omitempty"`
+}
+
+func (a Addon) IsEmpty() bool {
+	return reflect.DeepEqual(a, Addon{})
```
Comment (vdemeester) [inline pkg/apis/operator/v1alpha1/tektonaddon_types.go:78]: Any reason to use `reflect` here instead of just `len(a.Params) == 0` ?
We should avoid using `reflect` when we can, it's kind-of costly for.. not always a valid reason to be.

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 1 above)
Comment (vdemeester) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:32]: Note that if we change what is written in the Status (the name of the condition, …), we need to validate other tools that are depending on this won't be broken (e.g. openshift console, …)

Comment (vdemeester) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:32]: (because we are changing the API)

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle_test.go [hunk 9]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+    httc://www.apache.org/licenses/LICENSE-2.0
```
Comment (vdemeester) [inline pkg/apis/operator/v1alpha1/tektonconfig_lifecycle_test.go:8]: http

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go [hunk 10]
```diff
@@ -21,107 +21,128 @@ import (
...
-func (tp *TektonConfig) GroupVersionKind() schema.GroupVersionKind {
+func (tc *TektonConfig) GroupVersionKind() schema.GroupVersionKind {
 	return SchemeGroupVersion.WithKind(KindTektonConfig)
 }
 
-// GetCondition returns the current condition of a given condition type
-func (tps *TektonConfigStatus) GetCondition(t apis.ConditionType) *apis.Condition {
-	return configCondSet.Manage(tps).GetCondition(t)
+func (tc *TektonConfig) GetGroupVersionKind() schema.GroupVersionKind {
+	return SchemeGroupVersion.WithKind(KindTektonConfig)
```
Comment (vdemeester) [inline pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:41]: Any reason to have both ?

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go [hunk 11]
```diff
@@ -95,13 +90,13 @@ type TektonConfigSpec struct {
...
-	// The version of the installed release
+	// The profile installed
 	// +optional
-	Version string `json:"version,omitempty"`
+	Profile string `json:"profile,omitempty"`
 
-	// The url links of the manifests, separated by comma
+	// The version of the installed release
 	// +optional
-	Manifests []string `json:"manifests,omitempty"`
```
Comment (vdemeester) [inline pkg/apis/operator/v1alpha1/tektonconfig_types.go:104]: Are we breaking the status API by removing this ?

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 1 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:32]: yeah in this case we are chaning the api
we have removed the manifest field from status
and updated conditions
what tools do we have to validate? 🤔

File: pkg/apis/operator/v1alpha1/tektonaddon_types.go (hunk 8 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonaddon_types.go:78]: yeah len of param would work... 
@vdemeester how about using `equality.Semantic.DeepEqual` not in this case but [here](https://github.com/tektoncd/operator/blob/039009587fe6bd89f975455a98c39113530f097e/pkg/reconciler/shared/tektonconfig/pipeline/pipeline.go#L61) is that okay compared to reflect?

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 10 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:41]: `GroupVersionKind` can be removed  now.. will update !

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go (hunk 11 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:104]: yes

Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:104]: previously it was nil though.. we didn't use to add anything to that field in TektonConfig

File: pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go (hunk 10 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_lifecycle.go:41]: in generated code and internally in knative/pkg `` is used 
`GetGroupVersionKind` is defined so that we can do
```
tektonTriggerinformer.Get(ctx).Informer().AddEventHandler(cache.FilteringResourceEventHandler{
			FilterFunc: controller.FilterController(&v1alpha1.TektonConfig{}),
			Handler:    controller.HandleAll(impl.EnqueueControllerOf),
		})
```
so lets keep `GroupVersionKind` also 
i will take a deep look and then remove

File: pkg/apis/operator/v1alpha1/tektonaddon_types.go (hunk 8 above)
Comment (vdemeester) [reply pkg/apis/operator/v1alpha1/tektonaddon_types.go]: Maybe yes ? My point is just that we should aim to use as less as possible `reflect` and those things.. but when we don't have the choice .. 😛

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go (hunk 3 above)
Comment (sm43) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:31]: https://github.com/tektoncd/operator/issues/424
created an issue
